### PR TITLE
tests: socketpair: fix userspace thread permissions

### LINKED ISSUE #25270: fix userspace permissions in socketpair tests
**Is your enhancement proposal related to a problem? Please describe.**
This is a follow-up for PR #24367 to be completed during the stabilization period before 2.3.0 is released.

Several tests are disabled in [tests/net/socket/socketpair/src/main.c](https://github.com/zephyrproject-rtos/zephyr/blob/master/tests/net/socket/socketpair/src/main.c) due to incorrect userspace permissions, incorrect initialization of userspace k_objects, etc.

**Describe the solution you'd like**
I would like to submit another patch to correct it, hopefully with the help of @andrewboie (he seems to know a lot of the detail I am unfamiliar with).

**Describe alternatives you've considered**

**Additional context**

Currently, if I re-enable the tests and build for `BOARD=qemu_cortex_m3` or `native_posix_64`, It executes successfully with

```bash
$ grep -v ^# ~/.zephyrrc | sed -e '/^$/d'
export ZEPHYR_TOOLCHAIN_VARIANT=zephyr
export ZEPHYR_SDK_INSTALL_DIR=~/zephyr-sdk-0.11.2
export BOARD=qemu_cortex_m3
export ZEPHYR_PROJECT=tests/net/socket/socketpair
$ source zephyr-env.sh
$ west build -p always $ZEPHYR_PROJECT -t run
```
However, if I run the following sanity check, then the fun begins.
```bash
$ ./scripts/sanitycheck --enable-slow --inline-logs -j8 --testcase-root tests/net/socket/socketpair/
Renaming output directory to /home/cfriedt/workspace/zephyrproject/zephyr/sanity-out.8
INFO    - JOBS: 8
INFO    - Selecting default platforms per test case
INFO    - Building initial testcase list...
INFO    - 6 test configurations selected, 242 configurations discarded due to filters.
INFO    - Adding tasks to the queue...
INFO    - Total complete:    1/   6  16%  skipped:    0, failed:    0
ERROR   - mps2_an385                tests/net/socket/socketpair/net.socket.socketpair  FAILED: Timeout
INFO    - /home/cfriedt/workspace/zephyrproject/zephyr/sanity-out/mps2_an385/tests/net/socket/socketpair/net.socket.socketpair/handler.log
ERROR   - *** Booting Zephyr OS build v2.3.0-rc1-26-g4980a13fc4f7  ***
...
Running test suite socketpair_only_kernel
===================================================================
starting test - test_socketpair_write_block
E: 0x20000180 is not a valid k_thread
E: syscall z_vrfy_k_thread_create failed check: access denied
E: r0/a1:  0x00000000  r1/a2:  0x00000000  r2/a3:  0x00000000
E: r3/a4:  0x00000000 r12/ip:  0x00000000 r14/lr:  0x00000000
E:  xpsr:  0x00000000
E: Faulting instruction address (r15/pc): 0x00000000
E: >>> ZEPHYR FATAL ERROR 3: Kernel oops on CPU 0
E: Current thread: 0x2000126c (unknown)
E: Halting system

INFO    - /home/cfriedt/workspace/zephyrproject/zephyr/sanity-out/mps2_an385/tests/net/socket/socketpair/net.socket.socketpair/handler.log
INFO    - Total complete:    4/   6  66%  skipped:    0, failed:    1
ERROR   - qemu_x86                  tests/net/socket/socketpair/net.socket.socketpair  FAILED: Exited with 2
INFO    - /home/cfriedt/workspace/zephyrproject/zephyr/sanity-out/qemu_x86/tests/net/socket/socketpair/net.socket.socketpair/handler.log
ERROR   - SeaBIOS (version rel-1.12.1-0-ga5cab58-dirty-20200214_052440-f7294c49af13-zephyr
)
Booting from ROM..Optimal CONFIG_X86_MMU_PAGE_POOL_PAGES 11
*** Booting Zephyr OS build v2.3.0-rc1-26-g4980a13fc4f7  ***
...
Running test suite socketpair_only_kernel
===================================================================
starting test - test_socketpair_write_block
E: 0x001251a0 is not a valid k_thread
E: syscall z_vrfy_k_thread_create failed check: access denied
E: EAX: 0x00000000, EBX: 0x00000000, ECX: 0x00000000, EDX: 0x00000000
E: ESI: 0x00000000, EDI: 0x00000000, EBP: 0x00000000, ESP: 0x00161f84
E: EFLAGS: 0x00000246 CS: 0x002b CR3: 0x001361e0
E: call trace:
E: EIP: 0x001006a6
E: NULL base ptr
E: >>> ZEPHYR FATAL ERROR 3: Kernel oops on CPU 0
E: Current thread: 0x00136ea0 (unknown)
E: Halting system

INFO    - /home/cfriedt/workspace/zephyrproject/zephyr/sanity-out/qemu_x86/tests/net/socket/socketpair/net.socket.socketpair/handler.log
INFO    - Total complete:    6/   6  100%  skipped:    0, failed:    2
INFO    - 4 of 6 tests passed (66.67%), 2 failed, 0 skipped with 0 warnings in 89.14 seconds
INFO    - In total 1 test cases were executed on 6 out of total 249 platforms (2.41%)
INFO    - 4 tests executed on platforms, 2 tests were only built.
```

### PR DESCRIPTION
**Pull request by cfriedt**
Kernel objects were being directly accessed without previously calling k_thread_access_grant().
    
This change allows each test that requires an asynchronous event to send it to a common work queue with correct permissions given to the userspace thread.
    
Also, clarified some conditions in subsys/net/lib/sockets/socketpair.c and added / reordered some calls to respect whether semaphores were held / object were valid.
    
Fixes #25270

### CODE REVIEW

**Comment by cfriedt** [comment]
```
Does it make sense for k_thread_create() to just create a k_object of type K_OBJ_THREAD for the calling  struct k_thread that is the first parameter, if the call is from userspace?

Chris Friedt  5 hours ago
Z_OOPS(Z_SYSCALL_OBJ_NEVER_INIT(new_thread, K_OBJ_THREAD));
Is what's oopsing. new_thread is zero-initialized, but it's still oopsing. Not sure why.

Chris Friedt  5 hours ago
If I comment that out, I get either:
E: thread %p (%u) does not have permission on z_thread_stack element %p , and then if I manually expand the macro but change the section to ZTEST_DMEM I get E: %p is not a valid z_thread_stack_element . I'm sure that there is a simple fix for this sort of thing, but I have no idea what it is.

andrewboie  4 hours ago
ok there's no need to dynamically allocate anything.

andrewboie  4 hours ago
is this code already in-tree? it looks like your patch just removes an ifdef

andrewboie  4 hours ago
first problem with your code: ZTEST_BMEM struct k_thread th

andrewboie  4 hours ago
you can't do this with kernel objects.

andrewboie  4 hours ago
your k_thread can't be put in a user memory domain, that's why it is telling you it is invalid

andrewboie  4 hours ago
remove the ZTEST_BMEM

andrewboie  4 hours ago
and assign permissions instead.

andrewboie  4 hours ago
so what you are missing is 1) you are trying to put kernel objects in user memory and that doesn't work 2) you aren't assigning permissions on these kernel objects to your user threads so that they can access them

andrewboie  4 hours ago
$ git diff
diff --git a/tests/net/socket/socketpair/src/main.c b/tests/net/socket/socketpair/src/main.c
index 046f17b0d1..470212e194 100644
--- a/tests/net/socket/socketpair/src/main.c
+++ b/tests/net/socket/socketpair/src/main.c
@@ -39,10 +39,18 @@ extern void test_socketpair_poll_close_remote_end_POLLOUT(void);
 extern void test_socketpair_poll_immediate_data(void);
 extern void test_socketpair_poll_delayed_data(void);
+extern struct k_thread th;
+extern K_THREAD_STACK_DEFINE(th_stack, STACK_SIZE);
+
 void test_main(void)
 {
+
+       k_thread
+
        k_thread_system_pool_assign(k_current_get());
+       k_thread_access_grant(k_current_get(), &th, th_stack);
+
        ztest_test_suite(
                socketpair,
                ztest_user_unit_test(test_socketpair_AF_LOCAL__SOCK_STREAM__0),
@@ -70,13 +78,6 @@ void test_main(void)
        ztest_run_test_suite(socketpair);
-/* 20200509: @cfriedt: experiencing some issues with userspace thread
- * / memory permissions that will require some sorting out. Currently
- * these tests succeeed for native_posix_64.
- *
- * This feature is experimental at present.
- */
-#if 0
        ztest_test_suite(
                socketpair_only_kernel,
                ztest_user_unit_test(test_socketpair_write_block),
@@ -89,5 +90,4 @@ void test_main(void)
        );
        ztest_run_test_suite(socketpair_only_kernel);
-#endif
 }
diff --git a/tests/net/socket/socketpair/src/test_socketpair_block.c b/tests/net/socket/socketpair/src/test_socketpair_block.c
index b5a9cada0a..14f3c09d59 100644
--- a/tests/net/socket/socketpair/src/test_socketpair_block.c
+++ b/tests/net/socket/socketpair/src/test_socketpair_block.c
@@ -43,9 +43,9 @@ struct ctx {
 ZTEST_BMEM struct ctx ctx;
 #define STACK_SIZE 512
 /* thread structure for secondary thread */
-ZTEST_BMEM struct k_thread th;
+struct k_thread th;
 /* stack for the secondary thread */
-static K_THREAD_STACK_DEFINE(th_stack, STACK_SIZE);
+K_THREAD_STACK_DEFINE(th_stack, STACK_SIZE);
 static void th_fun(void *arg0, void *arg1, void *arg2)
 {

andrewboie  4 hours ago
something like that

andrewboie  4 hours ago
https://docs.zephyrproject.org/latest/reference/usermode/kernelobjects.html

andrewboie  4 hours ago
unfortuantely looks like socketpair has special requirements to get going, I can't just run it, otherwise I could just make the changes and send a PR...but hopefully this helps you get the idea




Chris Friedt  3 hours ago
Thanks @andrewboie!
```

**Comment by cfriedt** [comment]
Realized yesterday that I was reinventing the work queue. Should probably use one of those instead of manually doing it with a thread, stack, k_thread_suspend() / k_thread_resume(). Made a huge difference. I think only one of the tests is hanging atm, so will need to debug that a bit.

**Comment by cfriedt** [comment]
99.88% of tests passing but a page fault on qemu_x86. Any pointers on how to get rid of that?

**Comment by cfriedt** [comment]
Paging Dr. @andrewboie

**Comment by cfriedt** [comment]
I think my assumption was that a thread's stack was fair game from zephyr kernel space was wrong - at least on x86. I guess I need to use a `struct k_mem_slab`?

**Comment by andrewboie** [comment]
Were you going to post the contents of the error message? A page fault means you accessed memory improperly in some way.

What is going on at the PC value (EIP) when the fault happens? Have you attached a debugger and gotten a stack trace?
What memory location is it faulting on, what is located there?
What other steps have you taken to debug this?

**Comment by cfriedt** [comment]
error message:
```
SeaBIOS (version rel-1.12.1-0-ga5cab58-dirty-20200214_052440-f7294c49af13-zephyr
)
Booting from ROM..Optimal CONFIG_X86_MMU_PAGE_POOL_PAGES 11
*** Booting Zephyr OS build v2.3.0-rc1-98-g21dc7a957ddf  ***
Running test suite socketpair
===================================================================
starting test - test_socketpair_read_block
E: Page fault at address 0x15e004 (error code 0x4)
E: Linear address not present in page tables
E:   PDE: 0x000000000017d827 Writable, User, Execute Enabled
E:   PTE: Non-present
E: EAX: 0x0015e000, EBX: 0x00180033, ECX: 0x00000001, EDX: 0x00182f64
E: ESI: 0x00000000, EDI: 0x00182fe8, EBP: 0x00182ef0, ESP: 0x00182ef0
E: EFLAGS: 0x00000216 CS: 0x002b CR3: 0x0015e1e0
E: call trace:
E: EIP: 0x00132897
E:      0x0013313b (0x15e000)
E:      0x00100c95 (0x15e000)
E:      0x001013c8 (0x15e000)
E:      0x0012d8a0 (0x0)
E:      0x0012d911 (0x14d000)
E:      0x00103281 (0x14d000)
E: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
E: Current thread: 0x0015ee00 (unknown)
E: Halting system
```
backtrace:
```
#0  sys_sflist_peek_tail (list=0x15e000 <test_socketpair_work_q>) at ../include/sys/sflist.h:251
#1  0x0013313b in k_queue_append (queue=0x15e000 <test_socketpair_work_q>, data=0x182f5c <ztest_thread_stack+28508>) at /home/cfriedt/workspace/zephyrproject/zephyr/kernel/queue.c:193
#2  0x00100c95 in k_work_submit_to_queue (work_q=0x15e000 <test_socketpair_work_q>, work=0x182f5c <ztest_thread_stack+28508>) at /home/cfriedt/workspace/zephyrproject/zephyr/include/kernel.h:3128
#3  0x001013c8 in test_socketpair_read_block () at /home/cfriedt/workspace/zephyrproject/zephyr/tests/net/socket/socketpair/src/test_socketpair_block.c:150
#4  0x0012d8a0 in run_test_functions (test=0x14d000 <_socketpair.5261>) at /home/cfriedt/workspace/zephyrproject/zephyr/subsys/testsuite/ztest/src/ztest.c:177
#5  0x0012d911 in test_cb (a=0x14d000 <_socketpair.5261>, dummy2=0x0, dummy=0x0) at /home/cfriedt/workspace/zephyrproject/zephyr/subsys/testsuite/ztest/src/ztest.c:307
#6  0x00103281 in z_thread_entry (entry=0x12d8f0 <test_cb>, p1=0x14d000 <_socketpair.5261>, p2=0x0, p3=0x0) at /home/cfriedt/workspace/zephyrproject/zephyr/lib/os/thread_entry.c:29
#7  0x00000000 in ?? ()
```
```
(gdb) info sym 0x15e004
test_socketpair_work_q + 4 in section bss
(gdb) info line *((void *)0x00132897)
Line 251 of "../include/sys/sflist.h" starts at address 0x132894 <sys_sflist_peek_tail+3> and ends at 0x13289a <sys_sflist_peek_tail+9>.
```
Originally I thought maybe the issue was the user thread accessing the static "struct ctx" that I have in `tests/net/socket/socketpair/src/test_socketpair_block.c`, so then I tried to move the struct ctx to the thread stack, but it makes no difference. Also tried making it static with the ZTEST_BMEM attribute.

Trying to follow along with tests/kernel/workq/work_queue_api/src/main.c ..

**Comment by cfriedt** [comment]
Aha - `k_work_submit_to_user_queue()` was the missing link. Sorry for the spam @andrewboie !

File: tests/net/socket/socketpair/src/main.c [hunk 1]
```diff
@@ -31,63 +21,75 @@ extern void test_socketpair_unsupported_calls(void);
...
+static K_THREAD_STACK_DEFINE(test_socketpair_work_q_stack, 512);
+struct k_work_q test_socketpair_work_q;
 
 void test_main(void)
 {
 	k_thread_system_pool_assign(k_current_get());
 
+	k_thread_access_grant(k_current_get(), &test_socketpair_work_q.thread,
+	    &test_socketpair_work_q.queue,
+	    &test_socketpair_work_q_stack);
```
Comment (andrewboie) [inline tests/net/socket/socketpair/src/main.c:54]: I don't think you need to grant permission on the stack object in order for user threads to submit to the queue. The workqueue itself is being set up in supervisor mode (which requires no permissions), and the workqueue runs in supervisor mode since it is started with `k_work_q_start`. So if the only interaction with a user thread is to enqueue items the workqueue, I think permission on `&test_socketpair_work_q.queue` may be sufficient for descendant user threads.

**Review by tyrantlizard77 — DISMISSED** [review]
some minor comments otherwise LGTM

File: subsys/net/lib/sockets/socketpair.c [hunk 2]
```diff
@@ -385,6 +384,7 @@ static ssize_t spair_write(void *obj, const void *buffer, size_t count)
 		goto out;
 	}
 
+	/* slight race condition between k_sem_take and is_nonblock here */
```
[the remark this answers is not in the document]
Comment (cfriedt) [reply subsys/net/lib/sockets/socketpair.c:387]: In the very unlikely case that the scheduler swaps out this thread for another between `is_nonblock =..` and `res = k_sem_take..`, and another thread changes modifies the bit that sets `is_nonblock`, then the value of `is_nonblock` would be incorrect.
Not sure what the proper approach is to mitigating that (extremely unlikely) other than briefly disabling the scheduler for those two lines.

File: tests/net/socket/socketpair/src/main.c (hunk 1 above)
Comment (cfriedt) [reply tests/net/socket/socketpair/src/main.c:54]: I was following the example in [tests/kernel/workq](https://github.com/zephyrproject-rtos/zephyr/blob/master/tests/kernel/workq/work_queue_api/src/main.c#L867). I could try without though.

File: subsys/net/lib/sockets/socketpair.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (andrewboie) [reply subsys/net/lib/sockets/socketpair.c:387]: Extremely unlikely cases tend to happen *all the time* at sufficient scale. It's not an argument for dismissing the risk.

I think you should reconsider the design of `sock_is_nonblock`, or at least how you are using it, if this is the case.

File: tests/net/socket/socketpair/src/main.c (hunk 1 above)
Comment (cfriedt) [reply tests/net/socket/socketpair/src/main.c]: Yep, you were right ;-)

File: subsys/net/lib/sockets/socketpair.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (cfriedt) [reply subsys/net/lib/sockets/socketpair.c:387]: I don't think it's possible to redesign it in this case. If the risk must be mitigated, then I think the lock / check / unlock, take / re-check dance might be the only option.

[the remark this answers is not in the document]
Comment (cfriedt) [reply subsys/net/lib/sockets/socketpair.c]: It is done.

**Comment by cfriedt** [comment]
Testing seems to be kind of slow. Here is the output of sanitycheck on my machine locally.
```
$ ./scripts/sanitycheck --enable-slow --inline-logs -j8 --testcase-root tests/net/socket/socketpair/
Renaming output directory to /home/cfriedt/workspace/zephyrproject/zephyr/sanity-out.1
INFO    - JOBS: 8
INFO    - Selecting default platforms per test case
INFO    - Building initial testcase list...
INFO    - 6 test configurations selected, 242 configurations discarded due to filters.
INFO    - Adding tasks to the queue...
INFO    - Total complete:    6/   6  100%  skipped:    0, failed:    0
INFO    - 6 of 6 tests passed (100.00%), 0 failed, 0 skipped with 0 warnings in 101.16 seconds
INFO    - In total 1 test cases were executed on 6 out of total 249 platforms (2.41%)
INFO    - 5 tests executed on platforms, 1 tests were only built.
```

**Comment by cfriedt** [comment]
Looks like this is failing due to #25428

**Comment by cfriedt** [comment]
@pfalcon @jukkar @andrewboie - anyone able to approve?

**Comment by pfalcon** [comment]
Commit message:

> Also, clarified some conditions in subsys/net/lib/sockets/socketpair.c and added / reordered some calls to respect whether semaphores were held / object were valid.

My usual comment would be that the changes to subsys/net/lib/sockets/socketpair.c are substantial enough to go in a separate commit and not piggy-back on a commit, which patches tests, with a fineprint "Also, ..." at the end.

**Comment by carlescufi** [comment]
@pfalcon and @andrewboie would very much appreciate a review for this one

**Comment by andrewboie** [comment]
I really think your changes for is_nonblock synchronization need to be in a separate commit. You should not keep adding unrelated stuff to the same commit.

Ultimately the networking team needs to approve this. The changes to support user mode look fine.

**Comment by cfriedt** [comment]
Split into 2 commits. Ok to merge, @pfalcon @jukkar @andrewboie?

**Review by pfalcon — APPROVED** [review]
Thanks.

**Review by tyrantlizard77 — APPROVED** [review]
user mode changes look OK.
no opinion on the race condition change.

File: subsys/net/lib/sockets/socketpair.c [hunk 3]
```diff
@@ -401,17 +402,15 @@ static ssize_t spair_write(void *obj, const void *buffer, size_t count)
 			res = -1;
 			goto out;
 		}
+		is_nonblock = sock_is_nonblock(spair);
```
Comment (andrewboie) [inline subsys/net/lib/sockets/socketpair.c:405]: Others know this mechanism better, but what's stopping sock_is_nonblock() from changing below this line? I'm not sure I understand how this closes the race. You have no synchronization object to prevent the nonblocking state from changing by some other thread (either preemption or running on another CPU) or ISR (either preempting this thread or firing on another CPU, if this state can be changed from ISRs).

Now in this test case it might not matter, this is more of a question as to what is the intended use of this API to begin with, since it never gives you *any* guarantees that the answer it gives you won't change at any time?

In other words, I don't think your change here fixes anything and it would probably be best to just drop this patch and leave this alone, since at least for this test case there will not be other threads or ISRs running that change the nonblocking state.

Comment (cfriedt) [reply subsys/net/lib/sockets/socketpair.c:405]: I can understand why it is a bit confusing.

To answer your first question though, taking the semaphore on 399 prevents any other thread from changing the nonblocking state below 399 (and subsequently 405).

`spair->sem` is the synchronization object.

The confusing part is that you can't block on taking it the first time in case the socket is non-blocking. So if the socket is non-blocking, *and* you fail to take the semaphore, you know _with certainty_ that you must error and set EAGAIN.

Does that clarify the rest?

Comment (andrewboie) [reply subsys/net/lib/sockets/socketpair.c:405]: I  think I see it now. Thanks for the extra detail!

Comment (pfalcon) [reply subsys/net/lib/sockets/socketpair.c:405]: > Others know this mechanism better, but what's stopping sock_is_nonblock() from changing below this line?

> In other words, I don't think your change here fixes anything and it would probably be best to just drop this patch and leave this alone

A few words from me. I heard there's somethings scribbled on the stone "Thou shalt not share I/O resource among threads", so whoever does,  is looking for adventure. Of course, we should not open up security holes accessible with this, or crash badly (which is considered a security issue itself nowadays). I would conjecture that standard TOCTOU mitigation measure - make an important state copy at the beginning, consistently use the copy after wards - would help here. Oh, if we really talk of "mitigation", that copy apparently should be marked as volatile, given modern compilers' desire to eat more and more into the C semantics.

Beyond that, I neither tried to understand the changes in detail, I base my approval on the fact that we had pretty detailed discussion of various issues with @cfriedt, and there're pretty extensive tests, so it should be ok (and we'll anyway need a few iterations to get it right). So, I'm just trying to make sure that organizationally, changes are properly structured, e.g. code change can be reverted without affecting tests (the good criteria to consider what can (==should) go into separate commits).

> I neither tried to understand the changes in detail

(Oh, I don't do that because how would it look if I deeply nitpicked into somebody's code, given that this work brought up some known issues and technical debt in "my" code, and that's what I should be looking into instead? ;-) ).

Comment (pfalcon) [reply subsys/net/lib/sockets/socketpair.c:405]: > Oh, if we really talk of "mitigation", that copy apparently should be marked as volatile, given modern compilers' desire to eat more and more into the C semantics.

Btw @andrewboie, wonder if syscall vrfy handlers are safe in that regard ;-).

Comment (andrewboie) [reply subsys/net/lib/sockets/socketpair.c:405]: > Btw @andrewboie, wonder if syscall vrfy handlers are safe in that regard ;-).

I'm not sure I understand what you mean here.

Comment (pfalcon) [reply subsys/net/lib/sockets/socketpair.c:405]: > I'm not sure I understand what you mean here.

Well, roughly speaking, a vrfy handler with a structure like https://github.com/zephyrproject-rtos/zephyr/blob/master/kernel/msg_q.c#L92 . Suppose we'd want to validate something *inside* struct k_msgq *q. Per TOCTOU rules, reliable way to do that would be to make a copy first:

~~~
struct k_msgq s = *q;

validate(s.some_field);
impl(&s);
~~~

My concern is that oversmart modern compilers could optimize that to:

~~~
validate(q->some_field);
impl(q);
~~~

To guaranteedly preclude that, we'd need to do `volatile struct k_msgq s = *q;`. And I don't believe I ever saw `volatile` keyword in any vrfy function, hence my question.

Comment (andrewboie) [reply subsys/net/lib/sockets/socketpair.c:405]: In the provided code snippet you have no idea whether q is even valid memory or not. For the situation you describe the best practice is to use the `z_user_from_copy()` API.

I'm not sure why you're linking `z_vrfy_k_msgq_alloc_init` that's not an example of this situation.

I don't know if compilers perform the optimization you cite, but some evidence of your claims would be helpful. Or are you just speculating?

In any event `z_user_to_copy()` and `z_user_from_copy()` are what we want people using, there is no chance of that happening if they are used and they further will later enable us to implement security features like SMEP and SMAP.

There is discussion of these considerations here: https://docs.zephyrproject.org/latest/reference/usermode/syscalls.html#verification-memory-access-policies

Comment (pfalcon) [reply subsys/net/lib/sockets/socketpair.c:405]: > I'm not sure why you're linking z_vrfy_k_msgq_alloc_init that's not an example of this situation.

Just a result of a quick grep of a func which accept a struct in memory, I don't imply there's something wrong with that func.

> Or are you just speculating?

Yep, just speculating, given that we discussed a case of how we could protect from races of external state change behind our back.

> There is discussion of these considerations here: https://docs.zephyrproject.org/latest/reference/usermode/syscalls.html#verification-memory-access-policies

Thanks for responses, I don't think I did all the homework I wanted on userspace and syscalls (another reason why converting posix ops to syscall was still in backlog).
